pipeline/lib/metric.ml

```diff
@@ -0,0 +1,230 @@
+              ( Fmt.str "%s\n### Test: %s\n" entry test_name,
+                SeenBenchmarks.add test_key seen )
+          | true -> (entry, seen)
+        in
+        let change =
+          if Float.is_nan change
+          then Fmt.str "- %s is a new metric" metric_name
+          else Fmt.str "- %s changed by %f%%" metric_name change
+        in
+        let entry = Fmt.str "%s%s" entry change in
+        (entry :: acc, seen))
+      changes ([], SeenBenchmarks.empty)
+  in
+  String.concat "\n" (List.rev text_list)
+
+let commit_url ~repo_id ~commit =
+  let url = Fmt.str "https://github.com/%s/commit/%s" repo_id commit in
+  let short_sha = String.sub commit 0 7 in
+  Fmt.str "[%s](%s)" short_sha url
+
+let pull_url ~repo_id ~pull_number =
+  let url = Fmt.str "https://github.com/%s/pulls/%d" repo_id pull_number in
+  let pull = Fmt.str "#%d" pull_number in
+  Fmt.str "[%s](%s)" pull url
+
+let format_changes_message ~changes ~repository ~worker ~docker_image
+    ~main_commit =
+  let pull_number = Repository.pull_number repository |> Option.get in
+  let commit = Repository.commit_hash repository in
+  let repo_id = Repository.info repository in
+  let main_commit = Option.value ~default:"Unknown hash" main_commit in
+  let header =
+    Fmt.str
+      "%s (%s) changes the metrics as follows in comparison to `main` (%s) \
+       when running on `%s (%s)`:"
+      (pull_url ~repo_id ~pull_number)
+      (commit_url ~repo_id ~commit)
+      (commit_url ~repo_id ~commit:main_commit)
+      worker docker_image
+  in
+  Fmt.str "%s\n%s" header (format_changes ~changes)
+
+let post_github_comment ~api ~repository ~pull_number ~text =
+  let open Lwt.Infix in
+  let repo_owner = Repository.owner repository in
+  let repo_name = Repository.name repository in
+  let variables =
+    [
+      ("owner", `String repo_owner);
+      ("name", `String repo_name);
+      ("pullNumber", `Int pull_number);
+    ]
+  in
+  let query =
+    {|
+       query ($owner: String!, $name: String!, $pullNumber: Int!) {
+         repository(owner: $owner, name: $name) {
+           pullRequest(number: $pullNumber) {
+             id
+           }
+         }
+       }
+     |}
+  in
+  let ( / ) a b = Yojson.Safe.Util.member b a in
+  let prId =
+    Github.Api.exec_graphql api query ~variables >|= fun json ->
+    let open Yojson.Safe.Util in
+    let repo = json / "data" / "repository" in
+    let id = repo / "pullRequest" / "id" in
+    id |> to_string
+  in
+  let _ =
+    prId >|= fun id ->
+    let variables = [ ("subjectId", `String id); ("body", `String text) ] in
+    let mutation =
+      {|
+           mutation ($body: String!, $subjectId: ID!) {
+             addComment(input:{body: $body, subjectId: $subjectId}) {
+               commentEdge {
+                 node {
+                   body
+                 }
+               }
+             }
+           }
+         |}
+    in
+    Github.Api.exec_graphql api mutation ~variables >|= fun json ->
```

Comment:
One thing I'm afraid of with OCurrent is that side effects can trigger more than once (notably after a restart of the pipeline). This graphql mutation may or may not be concerned (because we already do some stuff to avoid this...) but in doubt, could you look at [`Current_cache`](https://ocurrent.github.io/ocurrent/current/Current_cache/index.html) in order to allow only one comment per PR? :) (... I really really don't want us to spam)

Reply:
I've tried to implement caching with `Current_cache` based on the docs, but I think it could use a careful look @art-w. I really appreciate your intent of not spamming and I'm fully onboard. :)